examples/text/plot_hashing_vs_dict_vectorizer.py

```diff
@@ -3,109 +3,343 @@
+
 t0 = time()
-hasher = FeatureHasher(n_features=n_features, input_type="string")
-X = hasher.transform(tokens(d) for d in raw_data)
+hasher = FeatureHasher(n_features=2**18, input_type="string")
+X = hasher.transform(tokenize(d) for d in raw_data)
 duration = time() - t0
-print("done in %fs at %0.3fMB/s" % (duration, data_size_mb / duration))
-print("Found %d unique terms" % n_nonzero_columns(X))
+dict_count_vectorizers["vectorizer"].append(
+    hasher.__class__.__name__ + "\non raw tokens"
+)
+dict_count_vectorizers["speed"].append(data_size_mb / duration)
+print(f"done in {duration:.3f}s at {data_size_mb / duration:.3f}MB/s")
+print(f"Found {n_nonzero_columns(X)} unique tokens")
+
+# %%
+# We now plot the speed of the above methods for vectorizing.
+
+import matplotlib.pyplot as plt
+
+fig, ax = plt.subplots(figsize=(12, 6))
+
+y_pos = np.arange(len(dict_count_vectorizers["vectorizer"]))
+ax.barh(y_pos, dict_count_vectorizers["speed"], align="center")
+ax.set_yticks(y_pos)
+ax.set_yticklabels(dict_count_vectorizers["vectorizer"])
+ax.invert_yaxis()
+_ = ax.set_xlabel("speed (MB/s)")
+
+# %%
+# In both cases :func:`~sklearn.feature_extraction.FeatureHasher` is
+# approximately twice as fast as
+# :func:`~sklearn.feature_extraction.DictVectorizer`. This is handy when dealing
+# with large amounts of data, with the downside of losing the invertibility of
+# the transformation, which in turn makes the interpretation of a model a more
+# complex task.
+#
+# Comparison with special purpose text vectorizers
+# ------------------------------------------------
+#
+# :func:`~sklearn.feature_extraction.text.CountVectorizer` accepts raw data as
+# it internally implements tokenization and occurrence counting. It is similar
+# to the :func:`~sklearn.feature_extraction.DictVectorizer` when used along with
+# the customized function `token_freqs` as done in the previous section. The
+# difference being that :func:`~sklearn.feature_extraction.text.CountVectorizer`
+# is more flexible. In particular it accepts various regex patterns through the
+# `token_pattern` parameter.
+
+from sklearn.feature_extraction.text import CountVectorizer
+
+t0 = time()
+vectorizer = CountVectorizer()
+vectorizer.fit_transform(raw_data)
+duration = time() - t0
+dict_count_vectorizers["vectorizer"].append(vectorizer.__class__.__name__)
+dict_count_vectorizers["speed"].append(data_size_mb / duration)
+print(f"done in {duration:.3f}s at {data_size_mb / duration:.3f}MB/s")
+print(f"Found {len(vectorizer.get_feature_names_out())} unique terms")
+
+# %%
+# We see that using the :func:`~sklearn.feature_extraction.text.CountVectorizer`
+# implementation is approximately twice as fast as using the
+# :func:`~sklearn.feature_extraction.DictVectorizer` along with the simple
+# function we defined for mapping the tokens. The reason is that
+# :func:`~sklearn.feature_extraction.text.CountVectorizer` produces a sparse
+# representation of the counts using `scipy.sparse.csr_matrix`.
```

Comment:
`DictVectorizer` also outputs a sparse CSR matrix, so this is not the reason. The reason is probably that the tokenization code is made more efficient by reusing a compiled regular expression for the full training set instead of creating once per document as done in our naive `tokenize` function defined at the beginning of the notebook.